File: app/MultiMarket/go/db/fleas/flog/flog.go
// Copyright 10-Sep-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

// Fleas log.
package flog

import (
	"github.com/dedeme/MultiMarket/data/logRow"
	"github.com/dedeme/golib/date"
	"github.com/dedeme/golib/file"
	"github.com/dedeme/golib/json"
	"path"
)

var fpath string
var fid string

// Auxiliar function
func write(log []*logRow.T) {
	var a []json.T
	for _, e := range log {
		a = append(a, e.ToJs())
	}
	file.WriteAll(fpath, json.Wa(a).String())
}

// Auxiliar function.
func read() []*logRow.T {
	var a []*logRow.T
	for _, e := range json.FromString(file.ReadAll(fpath)).Ra() {
		a = append(a, logRow.FromJs(e))
	}
	return a
}

// Initializes log.
//    parent: Parent directory of "Flog.tb".
func Initialize(parent string) {
	fpath = path.Join(parent, "Flog.tb")
	if !file.Exists(fpath) {
		write([]*logRow.T{})
	}
}

// Resets flog (Delete its contents), generates a new id and returns it.
func NewId() string {
	write([]*logRow.T{})
	fid = date.Now().Format("%T")
	return fid
}

// Returns 'true' if "id" is correct.
//    id: Log identifier control.
func Check(id string) bool {
	return id != "" && id == fid
}

// Stops flog.
//    id: Log identifier control. If 'id' is outdated this function do nothing.
func Stop(id string) {
	if id != "" && id == fid {
		fid = ""
	}
}

// Adds a normal message.
//    id : Log identifier control. If 'id' is outdated this function do nothing.
//    msg: Message to show.
func Info(id, msg string) {
	if id != "" && id == fid {
		es := read()
		write(append([]*logRow.T{logRow.Info(msg)}, es...))
	}
}

// Reads Log and returns an 'JSONized' Arr[logRow.T], ordered from after to
// before.
//    id : Log identifier control. If 'id' is outdated 'ok==false'.
func Read(id string) (log json.T, ok bool) {
	if id != "" && id == fid {
		log = json.FromString(file.ReadAll(fpath))
		ok = true
	}
	return
}

File: app/MMarket_old202009/js/src/data/flea/Flog.js
// Copyright 08-Mar-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

import LogRow from "../LogRow.js"; // eslint-disable-line

/**
    Fleas log.
**/
export default class Flog {
  /**
      @param {string} id
      @param {!Array<!LogRow>} entries
  **/
  constructor (id, entries) {
    this._id = id;
    this._entries = entries;
  }

  /**
      @return {string}
  **/
  get id () {
    return this._id;
  }

  /**
      @return {!Array<!LogRow>}
  **/
  get entries () {
    return this._entries;
  }

  /**
      @param {!Array<?>} js
      @return {!Flog}
  **/
  static fromJs (js) {
    const /** string */ id = js[0];
    const /** !Array<!LogRow> */ es = js[1].map(e => LogRow.fromJs(e));
    return new Flog(id, es);
  }
}
